File: sdks/local-rp/dart/lib/src/dns/dns.dart
// DNS TXT record parsing, pinning, and vouch verification -- mirrors
// `crates/liblinkkeys/src/dns.rs` / the Rust/Go/Java reference SDKs' own
// `dns` modules. This library performs no I/O itself; [DnsResolver] is the
// network seam (design doc, "Required Network Access": every SDK needs a
// DNS TXT lookup capability, configurable, defaulting to the system
// resolver).
library;

import '../crypto/crypto.dart';
import '../errors.dart';
import '../local_rp.dart';
import '../wire/cbor.dart';
import '../wire/types.dart';

/// Default TCP port for the LinkKeys protocol service. Advertised `tcp=`
/// values omit the port when it equals this.
const int defaultTcpPort = 4987;

String linkkeysDnsName(String domain) => '_linkkeys.$domain';

String linkkeysApisDnsName(String domain) => '_linkkeys_apis.$domain';

/// A parsed `_linkkeys.{domain}` TXT record -- the trust anchor.
class LinkKeysRecord {
  final List<String> fingerprints;
  const LinkKeysRecord(this.fingerprints);
}

/// A parsed `_linkkeys_apis.{domain}` TXT record -- service endpoints.
class LinkKeysApis {
  final String? tcp;
  final String? httpsBase;
  const LinkKeysApis({this.tcp, this.httpsBase});
}

List<String> _fields(String txt) =>
    txt.split(RegExp(r'\s+')).where((p) => p.isNotEmpty).toList();

void _requireLk1Version(List<String> parts) {
  String? version;
  var found = false;
  for (final p in parts) {
    if (p.startsWith('v=')) {
      version = p.substring(2);
      found = true;
      break;
    }
  }
  if (!found) {
    throw DnsParseError(DnsParseErrorKind.missingVersion, null);
  }
  if (version != 'lk1') {
    throw DnsParseError(DnsParseErrorKind.unsupportedVersion, version);
  }
}

/// Parse a single `_linkkeys` TXT record string. Errors if it isn't a
/// LinkKeys v1 record.
LinkKeysRecord parseLinkKeysTxt(String txt) {
  final parts = _fields(txt);
  _requireLk1Version(parts);
  final fingerprints = <String>[];
  for (final p in parts) {
    if (p.startsWith('fp=')) {
      fingerprints.add(p.substring(3));
    }
  }
  return LinkKeysRecord(fingerprints);
}

String _normalizeTcpEndpoint(String value) {
  if (value.isEmpty || value.contains(':')) return value;
  return '$value:$defaultTcpPort';
}

/// Parse a single `_linkkeys_apis` TXT record string. Errors if it isn't a
/// LinkKeys v1 record or carries no endpoint.
LinkKeysApis parseLinkKeysApisTxt(String txt) {
  final parts = _fields(txt);
  _requireLk1Version(parts);

  String? tcp;
  String? httpsBase;
  for (final p in parts) {
    if (tcp == null && p.startsWith('tcp=')) {
      final v = _normalizeTcpEndpoint(p.substring(4));
      if (v.isNotEmpty) tcp = v;
    }
    if (httpsBase == null && p.startsWith('https=')) {
      final v = p.substring(6);
      if (v.isNotEmpty) httpsBase = 'https://$v';
    }
  }
  if (tcp == null && httpsBase == null) {
    throw DnsParseError(DnsParseErrorKind.missingApisEndpoint, null);
  }
  return LinkKeysApis(tcp: tcp, httpsBase: httpsBase);
}

/// Whether [fp] is a syntactically valid key fingerprint: 64 hex chars (a
/// SHA-256 digest), case-insensitive.
bool isValidFingerprint(String fp) {
  if (fp.length != 64) return false;
  for (var i = 0; i < fp.length; i++) {
    final c = fp.codeUnitAt(i);
    final hex = (c >= 0x30 && c <= 0x39) ||
        (c >= 0x61 && c <= 0x66) ||
        (c >= 0x41 && c <= 0x46);
    if (!hex) return false;
  }
  return true;
}

/// Pin fetched keys to the DNS-published fingerprint set: for each
/// candidate key, RECOMPUTE `fingerprint(public_key)` (never trust the wire
/// `fingerprint` field, which is attacker-controlled) and keep only keys
/// whose recomputed fingerprint is a member of [pinned].
Future<List<DomainPublicKey>> pinKeysToFingerprints(
    List<DomainPublicKey> keys, List<String> pinned) async {
  final pinnedLower = <String>{
    for (final f in pinned)
      if (isValidFingerprint(f)) f.toLowerCase(),
  };
  final out = <DomainPublicKey>[];
  for (final k in keys) {
    final fp = (await Crypto.fingerprint(k.publicKey)).toLowerCase();
    if (pinnedLower.contains(fp)) out.add(k);
  }
  return out;
}

const String _keyVouchTag = 'linkkeys-key-vouch-v1';

List<int> _keyVouchPayload(String encFingerprint, String encExpiresAt) {
  return Cbor.encode(Cbor.tuple([
    Cbor.vtext(_keyVouchTag),
    Cbor.vtext(encFingerprint),
    Cbor.vtext(encExpiresAt),
  ]));
}

/// Verify that [signingKey] vouches for [encKey]: the encryption key names
/// this signing key, the signing key is itself valid, and its signature
/// covers the recomputed encrypt-key fingerprint + expiry.
Future<bool> verifyKeyVouch(
    DomainPublicKey encKey, DomainPublicKey signingKey) async {
  if (encKey.signedByKeyId == null ||
      encKey.signedByKeyId != signingKey.keyId) {
    return false;
  }
  try {
    LocalRp.checkSigningKeyValid(signingKey);
  } catch (_) {
    return false;
  }
  if (encKey.keySignature == null) return false;
  if (signingKey.algorithm != 'ed25519') return false;
  final recomputedFp = await Crypto.fingerprint(encKey.publicKey);
  final payload = _keyVouchPayload(recomputedFp, encKey.expiresAt);
  return Crypto.verifyEd25519(
      payload, encKey.keySignature!, signingKey.publicKey);
}

/// Establish the trusted key set from a fetched key list and the DNS-pinned
/// fingerprint set: signing keys (`key_usage == "sign"`) are pinned
/// directly; encryption keys (`key_usage == "encrypt"`) are trusted only
/// when a DNS-pinned signing key vouches for them.
Future<List<DomainPublicKey>> trustKeys(
    List<DomainPublicKey> keys, List<String> pinned) async {
  final signing = keys.where((k) => k.keyUsage == 'sign').toList();
  final pinnedSigning = await pinKeysToFingerprints(signing, pinned);

  final trusted = <DomainPublicKey>[...pinnedSigning];
  for (final k in keys) {
    if (k.keyUsage != 'encrypt') continue;
    for (final sk in pinnedSigning) {
      if (await verifyKeyVouch(k, sk)) {
        trusted.add(k);
        break;
      }
    }
  }
  return trusted;
}

File: sdks/local-rp/java/src/main/java/community/catalyst/linkkeys/localrp/dns/Dns.java
package community.catalyst.linkkeys.localrp.dns;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Locale;
import java.util.Set;

import community.catalyst.linkkeys.localrp.crypto.Crypto;
import community.catalyst.linkkeys.localrp.wire.Cbor;
import community.catalyst.linkkeys.localrp.wire.Types.DomainPublicKey;

/**
 * DNS TXT record parsing, pinning, and vouch verification &mdash; mirrors
 * {@code crates/liblinkkeys/src/dns.rs} / {@code sdks/local-rp/go/dns.go}.
 * This class performs no I/O itself; {@link DnsResolver} is the network
 * seam (design doc, "Required Network Access": every SDK needs a DNS TXT
 * lookup capability, configurable, defaulting to the system resolver).
 */
public final class Dns {
    private Dns() {}

    /** Default TCP port for the LinkKeys protocol service. Advertised {@code tcp=} values omit the port when it equals this. */
    public static final int DEFAULT_TCP_PORT = 4987;

    public static String linkkeysDnsName(String domain) {
        return "_linkkeys." + domain;
    }

    public static String linkkeysApisDnsName(String domain) {
        return "_linkkeys_apis." + domain;
    }

    /** A parsed {@code _linkkeys.{domain}} TXT record &mdash; the trust anchor. */
    public record LinkKeysRecord(List<String> fingerprints) {}

    /** A parsed {@code _linkkeys_apis.{domain}} TXT record &mdash; service endpoints. */
    public record LinkKeysApis(String tcp, String httpsBase) {}

    private static String requireLk1Version(List<String> parts) {
        String version = null;
        boolean found = false;
        for (String p : parts) {
            if (p.startsWith("v=")) {
                version = p.substring(2);
                found = true;
                break;
            }
        }
        if (!found) {
            throw new DnsParseError(DnsParseError.Kind.MISSING_VERSION, null);
        }
        if (!"lk1".equals(version)) {
            throw new DnsParseError(DnsParseError.Kind.UNSUPPORTED_VERSION, version);
        }
        return version;
    }

    private static List<String> fields(String txt) {
        List<String> out = new ArrayList<>();
        for (String p : txt.split("\\s+")) {
            if (!p.isEmpty()) {
                out.add(p);
            }
        }
        return out;
    }

    /** Parse a single {@code _linkkeys} TXT record string. Errors if it isn't a LinkKeys v1 record. */
    public static LinkKeysRecord parseLinkKeysTxt(String txt) {
        List<String> parts = fields(txt);
        requireLk1Version(parts);
        List<String> fingerprints = new ArrayList<>();
        for (String p : parts) {
            if (p.startsWith("fp=")) {
                fingerprints.add(p.substring(3));
            }
        }
        return new LinkKeysRecord(fingerprints);
    }

    private static String normalizeTcpEndpoint(String value) {
        if (value.isEmpty() || value.contains(":")) {
            return value;
        }
        return value + ":" + DEFAULT_TCP_PORT;
    }

    /** Parse a single {@code _linkkeys_apis} TXT record string. Errors if it isn't a LinkKeys v1 record or carries no endpoint. */
    public static LinkKeysApis parseLinkKeysApisTxt(String txt) {
        List<String> parts = fields(txt);
        requireLk1Version(parts);

        String tcp = null;
        String httpsBase = null;
        for (String p : parts) {
            if (tcp == null && p.startsWith("tcp=")) {
                String v = normalizeTcpEndpoint(p.substring(4));
                if (!v.isEmpty()) {
                    tcp = v;
                }
            }
            if (httpsBase == null && p.startsWith("https=")) {
                String v = p.substring(6);
                if (!v.isEmpty()) {
                    httpsBase = "https://" + v;
                }
            }
        }
        if (tcp == null && httpsBase == null) {
            throw new DnsParseError(DnsParseError.Kind.MISSING_APIS_ENDPOINT, null);
        }
        return new LinkKeysApis(tcp, httpsBase);
    }

    /** Whether {@code fp} is a syntactically valid key fingerprint: 64 hex chars (a SHA-256 digest), case-insensitive. */
    public static boolean isValidFingerprint(String fp) {
        if (fp.length() != 64) {
            return false;
        }
        for (int i = 0; i < fp.length(); i++) {
            char c = fp.charAt(i);
            boolean hex = (c >= '0' && c <= '9') || (c >= 'a' && c <= 'f') || (c >= 'A' && c <= 'F');
            if (!hex) {
                return false;
            }
        }
        return true;
    }

    /**
     * Pin fetched keys to the DNS-published fingerprint set: for each
     * candidate key, RECOMPUTE {@code fingerprint(public_key)} (never trust
     * the wire {@code fingerprint} field, which is attacker-controlled) and
     * keep only keys whose recomputed fingerprint is a member of
     * {@code pinned}.
     */
    public static List<DomainPublicKey> pinKeysToFingerprints(List<DomainPublicKey> keys, List<String> pinned) {
        Set<String> pinnedLower = new HashSet<>();
        for (String f : pinned) {
            if (isValidFingerprint(f)) {
                pinnedLower.add(f.toLowerCase(Locale.ROOT));
            }
        }
        List<DomainPublicKey> out = new ArrayList<>();
        for (DomainPublicKey k : keys) {
            String fp = Crypto.fingerprint(k.publicKey()).toLowerCase(Locale.ROOT);
            if (pinnedLower.contains(fp)) {
                out.add(k);
            }
        }
        return out;
    }

    private static final String KEY_VOUCH_TAG = "linkkeys-key-vouch-v1alpha";

    static byte[] keyVouchPayload(String encFingerprint, String encExpiresAt) {
        return Cbor.encode(Cbor.tuple(Cbor.vtext(KEY_VOUCH_TAG), Cbor.vtext(encFingerprint), Cbor.vtext(encExpiresAt)));
    }

    /**
     * Verify that {@code signingKey} vouches for {@code encKey}: the
     * encryption key names this signing key, the signing key is itself
     * valid, and its signature covers the recomputed encrypt-key
     * fingerprint + expiry.
     */
    public static boolean verifyKeyVouch(DomainPublicKey encKey, DomainPublicKey signingKey) {
        if (encKey.signedByKeyId() == null || !encKey.signedByKeyId().equals(signingKey.keyId())) {
            return false;
        }
        try {
            community.catalyst.linkkeys.localrp.LocalRp.checkSigningKeyValid(signingKey);
        } catch (RuntimeException e) {
            return false;
        }
        if (encKey.keySignature() == null) {
            return false;
        }
        if (!"ed25519".equals(signingKey.algorithm())) {
            return false;
        }
        String recomputedFp = Crypto.fingerprint(encKey.publicKey());
        byte[] payload = keyVouchPayload(recomputedFp, encKey.expiresAt());
        return Crypto.verifyEd25519(payload, encKey.keySignature(), signingKey.publicKey());
    }

    /**
     * Establish the trusted key set from a fetched key list and the
     * DNS-pinned fingerprint set: signing keys ({@code key_usage == "sign"})
     * are pinned directly; encryption keys ({@code key_usage == "encrypt"})
     * are trusted only when a DNS-pinned signing key vouches for them.
     */
    public static List<DomainPublicKey> trustKeys(List<DomainPublicKey> keys, List<String> pinned) {
        List<DomainPublicKey> signing = new ArrayList<>();
        for (DomainPublicKey k : keys) {
            if ("sign".equals(k.keyUsage())) {
                signing.add(k);
            }
        }
        List<DomainPublicKey> pinnedSigning = pinKeysToFingerprints(signing, pinned);

        List<DomainPublicKey> trusted = new ArrayList<>(pinnedSigning);
        for (DomainPublicKey k : keys) {
            if (!"encrypt".equals(k.keyUsage())) {
                continue;
            }
            for (DomainPublicKey sk : pinnedSigning) {
                if (verifyKeyVouch(k, sk)) {
                    trusted.add(k);
                    break;
                }
            }
        }
        return trusted;
    }
}
